utils/serde/src/main/java/io/apicurio/registry/utils/serde/JsonSchemaKafkaDeserializer.java

```diff
@@ -0,0 +1,217 @@
+     */
+    public JsonSchemaKafkaDeserializer(RegistryService client) {
+        super(client);
+    }
+
+    /**
+     * @see org.apache.kafka.common.serialization.Deserializer#configure(java.util.Map, boolean)
+     */
+    @Override
+    public void configure(Map<String, ?> configs, boolean isKey) {
+        super.configure(configs);
+
+        Object ve = configs.get(REGISTRY_JSON_SCHEMA_DESERIALIZER_VALIDATION_ENABLED);
+        this.validationEnabled = Utils.isTrue(ve);
+
+        this.schemaCache = new SchemaCache<SchemaValidator>(getClient()) {
+            @Override
+            protected SchemaValidator toSchema(Response response) {
+                String schema = response.readEntity(String.class);
+                return api.loadSchema(new StringSchemaSource(schema));
+            }
+        };
+    }
+    
+    /**
+     * @see org.apache.kafka.common.serialization.Deserializer#deserialize(java.lang.String, byte[])
+     */
+    @Override
+    public T deserialize(String topic, byte[] data) {
+        throw new UnsupportedOperationException();
+    }
+    
+    /**
+     * @see org.apache.kafka.common.serialization.Deserializer#deserialize(java.lang.String, org.apache.kafka.common.header.Headers, byte[])
+     */
+    @Override
+    public T deserialize(String topic, Headers headers, byte[] data) {
+        if (data == null) {
+            return null;
+        }
+        
+        try {
+            JsonParser parser = mapper.getFactory().createParser(data);
+            if (validationEnabled) {
+                Long globalId = getGlobalId(headers);
+                
+                // If no globalId is provided, check the alternative - which is to check for artifactId and 
+                // (optionally) version.  If these are found, then convert that info to globalId.
+                if (globalId == null) {
+                    String artifactId = getArtifactId(headers);
+                    Integer version = getVersion(headers);
+                    if (version == null) {
+                        version = getLatestVersion(artifactId);
+                    }
+                    globalId = toGlobalId(artifactId, version);
+                }
+                
+                SchemaValidator schema = schemaCache.getSchema(globalId);
+                parser = api.decorateJsonParser(schema, parser);
+            }
+            
+            Class<T> messageType = getMessageType(headers);
+
+            return mapper.readValue(parser, messageType);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    /**
+     * Gets the global id from the headers.  Returns null if not found.
+     * @param headers
+     */
+    protected Long getGlobalId(Headers headers) {
+        Header header = headers.lastHeader(JsonSchemaSerDeConstants.HEADER_GLOBAL_ID);
+        if (header == null) {
+            return null;
+        }
+        ByteBuffer buffer = ByteBuffer.allocate(8);
```

Comment:
Just use ByteBuffer.allocate(header.value()) ?

Reply:
I assume you mean this:  `ByteBuffer.allocate(header.value().length)` which is fine.  I can change that.

Reply:
Eh, missed it again ... I meant ByteBuffer.wrap()